CONTRIBUTING.md

```diff
@@ -96,6 +96,98 @@ Every commit in your pull request needs to be able to build and pass the CI test
 
 If the pull request closes an issue please note it as: `"Fixes #NNN"`.
 
+## Unit Testing Guidelines
+
+### Unit Test Checks
+
+* The "testify" package should not be used.
+* The "cmp" package is allowed.
+* Unit tests in Go should follow the guidelines in this tutorial: https://go.dev/doc/tutorial/add-a-test
+  * In particular, the test error should be in the form `Function(...) = ...; want ...`.
+
+For example:
+
+```
+if msg != "" || err == nil {
+	t.Fatalf(`Hello("") = %q, %v, want "", error`, msg, err)
+}
+```
+
+### Mocking Dependencies
+
+* Special mocking packages should not be used.
+* Prefer to test the real thing where possible.
+* Mocking is sometimes necessary. For example:
+   * Operations as root.
+   * Interacting with special hardware (ex: USB, SPI, PCIe)
+   * Modifying machine state (ex: reboot, kexec)
+   * Tests which would otherwise be flaky (ex: `time.Sleep`, networking)
+* Consider writing an integration test for the above.
+* Prefer to mock using existing interfaces. For example: `io.Reader`, `fs.FS`
+* Avoid mocking using global state. Instead, consider using one of the
+  following "dependency injection" idioms:
+
+1. Mocking functions:
+
+```
+// The exported function has a meaningful API.
+func SetMemAddr(addr, val uint) error {
+	return setMemAddr(addr, val, "/dev/mem")
+}
+
+// The internal function is called from the unit test. The test can set a
+// different `file` argument.
+func setMemAddr(addr, val uint, file string) error {
+	f, err := os.Open(file)
+	if err != nil {
+		return err
+	}
+	...
+}
+```
+
+2. Mocking objects:
```

Comment:
need to fix this as it is a dup of 'Mocking functions'

Reply:
This is an oldie but goodie, maybe worth referring to for file system mocks
https://talks.golang.org/2012/10things.slide#8

Reply:
Done.

Reply:
The new way to mock out filesystem is https://pkg.go.dev/io/fs and https://pkg.go.dev/testing/fstest. I put a mention above.